pkg/lib/tmp/copy.go

```diff
@@ -0,0 +1,49 @@
+package tmp
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+func CopyTmpDB(src string) (tmpName string, err error) {
+	tmp, err := ioutil.TempFile("", "db-")
+	if err != nil {
+		return
+	}
+	tmpName = tmp.Name()
+	err = CopyRegularFile(src, tmp.Name())
+	if err != nil {
+		err = utilerrors.NewAggregate([]error{err, os.Remove(tmpName)})
+	}
+	return
+}
+
+func CopyRegularFile(srcName, dstName string) (err error) {
+	srcStat, err := os.Stat(srcName)
+	if err != nil {
+		return err
+	}
+	if !srcStat.Mode().IsRegular() {
+		return fmt.Errorf("%s is not a regular file", srcName)
+	}
+	src, err := os.Open(srcName)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = utilerrors.NewAggregate([]error{err, src.Close()})
+	}()
+	dst, err := os.Create(dstName)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = utilerrors.NewAggregate([]error{err, dst.Close()})
+	}()
+	_, err = io.Copy(dst, src)
```

Comment:
The one thing I'm thinking about is how much did we increase the vulnerability surface of the registry by running it as a user who has the ability to copy/manipulate random files. It may have been like this previously, but before we were only modifying one file during migration (the db) versus creating and copying files around. Maybe it's worth setting the name of the tempDB in the Dockerfile as an env argument and locking down the user permissions to only be able to modify those two files? Just a thought

Reply:
> running it as a user who has the ability to copy/manipulate random files.

To clarify, this is what the user running the server must have:

- read access to the database file
- write access to a new /tmp file

and /tmp is typically 777 + sticky bit.